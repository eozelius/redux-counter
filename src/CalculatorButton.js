import React from 'react'

const CalculatorButton = (props) => {
  return (
    <div className='calcBtn' onClick={props.onClick}>
      {props.display}
    </div>
  )
}

export default CalculatorButton
