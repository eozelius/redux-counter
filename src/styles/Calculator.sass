.calculatorContainer
  width: 33%
  margin: 1% auto 5%
  display: block

.displayContainer
  width: 100%
  height: 100px
  border: 1px solid teal
  display: block

.numberPadContainer
  width: 75%
  float: left
  display: inline-block

.operatorContainer
  width: 25%
  display: inline-block

.calcBtn
  width: 100px
  height: 100px
  border: 1px solid red
  display: inline-block
  flex-wrap: wrap
  line-height: 100px
  cursor: pointer
  font-size: 2em
  margin: 5px
