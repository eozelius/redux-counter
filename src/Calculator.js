import React, { Component } from 'react'
import './styles/Calculator.sass'
import NumberPad from './NumberPad'
import CalculatorDisplay from './CalculatorDisplay'
import OperatorPad from './OperatorPad'
import CalculatorButton from './CalculatorButton'

class Calculator extends Component {
  constructor(props){
    super(props)
    this.state = {
      firstNum: '',
      secondNum: '',
      operator: null,
      displayNum: '',
      complete: false
    }
  }

  handleNumber = (e) => {
    if(this.state.complete){
      const number = `${e.target.innerText}`
      this.setState({
        displayNum: number,
        firstNum: number,
        complete: false
      })
      return
    }

    // user is inputing X
    if(this.state.operator === null){
      const number = `${this.state.displayNum}${e.target.innerText}`
      this.setState({
        displayNum: number,
        firstNum: number
      })
    // User is inputing Y
    } else {
      const secondNum = `${this.state.secondNum}${e.target.innerText}`
      this.setState({
        secondNum: secondNum,
        displayNum: secondNum
      })
    }
  }

  handleOperator = (e) => {
    this.setState({
      operator: e.target.innerText
    })
  }

  calculate = () => {
    const x = parseFloat(this.state.firstNum)
    const y = parseFloat(this.state.secondNum)
    const op = this.state.operator
    const result = calculateResult(x, y, op)

    this.setState({
      displayNum: `${result}`,
      firstNum: '',
      secondNum: '',
      operator: null,
      complete: true
    })
  }

  render(){
    return(
      <div className='calculatorContainer' >
        <CalculatorDisplay number={this.state.displayNum} />
        <CalculatorButton className='executeResult' onClick={this.calculate} display='=' />
        <NumberPad handleNumber={this.handleNumber} />
        <OperatorPad handleOperator={this.handleOperator} />
      </div>
    )
  }
}

export function calculateResult(x, y, op){
  switch(op){
    case '+':
      return x + y
    case '-':
      return x - y
    case '/':
      return x / y
    case '*':
      return x * y
    default:
      // throw('operator must be "+", "-", "/" or "*"')
      return false
  }
}

export default Calculator
