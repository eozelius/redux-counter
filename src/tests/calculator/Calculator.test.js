import Calculator from '../../Calculator'
import React from 'react'
import CalculatorButton from '../../CalculatorButton'

describe(Calculator, () => {
  test('mounts properly', () => {
    const calc = shallow(<Calculator />)
    expect(calc).toMatchSnapshot()
  })

  test('initial state', () => {
    const expectedState = {
      firstNum: '',
      secondNum: '',
      operator: null,
      displayNum: '',
      complete: false
    }
    const calc = shallow(<Calculator />)
    expect(calc.state()).toEqual(expectedState)
  })

  test('it handles number input', () => {
    const calc = mount(<Calculator />)
    const nine = calc.find(CalculatorButton).at(2)
    nine.simulate('click', {
      target: {
        innerText: nine.prop('display')
      }
    })
    expect(calc.state('firstNum')).toEqual(nine.prop('display').toString())
  })

  test('it handles operator input', () => {
    const calc = mount(<Calculator />)
    const divide = calc.find('.operatorContainer .calcBtn').last()
    divide.simulate('click', {
      target: {
        innerText: divide.prop('children')
      }
    })
    expect(calc.state('operator')).toEqual(divide.prop('children').toString())
  })

  test('handles Number input', () => {
    const spy = jest.spyOn(Calculator.prototype, 'handleNumber')
    const calc = mount(<Calculator />)
    const nine = calc.find('.calcBtn').at(1)
    nine.simulate('click', {target: { innerText: nine.prop('display') }})
    expect(spy).toHaveBeenCalled()
  })

  test('it calculates result', () => {
    const calc = mount(<Calculator />)
    const nine = calc.find(CalculatorButton).at(1)
    const divide = calc.find('.operatorContainer .calcBtn').last()
    const equal = calc.find('.executeResult').first()
    nine.simulate('click', {target: { innerText: nine.prop('display') }})
    divide.simulate('click', {target: {innerText: divide.prop('children')}})
    nine.simulate('click', {target: { innerText: nine.prop('display') }})
    equal.simulate('click')
    expect(calc.state('displayNum')).toEqual('1')
    expect(calc.state('complete')).toEqual(true)
  })

  test('it resets state after "=" button pressed', () => {
    const calc = mount(<Calculator />)
    calc.setState({
      complete: true
    })
    const nine = calc.find(CalculatorButton).at(1)
    const expectedState = {
      firstNum: '9',
      secondNum: '',
      operator: null,
      displayNum: '9',
      complete: false
    }
    nine.simulate('click', {target: { innerText: nine.prop('display') }})
    expect(calc.state()).toEqual(expectedState)
  })
})
