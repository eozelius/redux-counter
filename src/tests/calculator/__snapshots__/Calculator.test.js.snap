// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator mounts properly 1`] = `
<div
  className="calculatorContainer"
>
  <CalculatorDisplay
    number=""
  />
  <CalculatorButton
    className="executeResult"
    display="="
    onClick={[Function]}
  />
  <NumberPad
    handleNumber={[Function]}
  />
  <OperatorPad
    handleOperator={[Function]}
  />
</div>
`;
